File: string-transformer/typescript/src/pipeline.ts
// Byte-identical across C#, TypeScript, and Python. The truncation
// marker is part of the spec (see ../SCENARIOS.md).
export const TRUNCATION_MARKER = '...';

// Strategy. Every transformation is a self-contained rule — one input,
// one output, no state. Pipeline composes them; it does not know which
// rule any given strategy implements.
export interface Transformation {
  apply(input: string): string;
}

export class Pipeline {
  constructor(private readonly steps: readonly Transformation[]) {}

  run(input: string): string {
    return this.steps.reduce((current, step) => step.apply(current), input);
  }
}

const isLetter = (c: string): boolean => /^[A-Za-z]$/.test(c);
const isSeparator = (c: string): boolean => /\s/.test(c) || c === '-';

export class Capitalise implements Transformation {
  apply(input: string): string {
    if (input.length === 0) return input;
    const chars = [...input];
    let atWordStart = true;
    for (let i = 0; i < chars.length; i++) {
      const c = chars[i]!;
      const letter = isLetter(c);
      if (letter && atWordStart) chars[i] = c.toUpperCase();
      atWordStart = !letter;
    }
    return chars.join('');
  }
}

export class Reverse implements Transformation {
  apply(input: string): string {
    return [...input].reverse().join('');
  }
}

export class RemoveWhitespace implements Transformation {
  apply(input: string): string {
    return input.replace(/\s+/g, '');
  }
}

export class SnakeCase implements Transformation {
  apply(input: string): string {
    const lowered = input.toLowerCase();
    let out = '';
    let inSeparator = true; // suppress leading separators
    for (const c of lowered) {
      if (isSeparator(c)) {
        inSeparator = true;
      } else {
        if (inSeparator && out.length > 0) out += '_';
        out += c;
        inSeparator = false;
      }
    }
    return out;
  }
}

export class CamelCase implements Transformation {
  apply(input: string): string {
    let out = '';
    let tokenIndex = 0;
    let atTokenStart = true;
    for (const c of input) {
      if (isSeparator(c)) {
        if (!atTokenStart) {
          tokenIndex++;
          atTokenStart = true;
        }
      } else {
        const lower = c.toLowerCase();
        out += atTokenStart && tokenIndex > 0 ? lower.toUpperCase() : lower;
        atTokenStart = false;
      }
    }
    return out;
  }
}

export class Truncate implements Transformation {
  constructor(private readonly length: number) {}

  apply(input: string): string {
    if (input.length <= this.length) return input;
    return input.slice(0, this.length) + TRUNCATION_MARKER;
  }
}

export class Repeat implements Transformation {
  constructor(private readonly times: number) {}

  apply(input: string): string {
    if (this.times <= 0) return '';
    return Array(this.times).fill(input).join(' ');
  }
}

export class Replace implements Transformation {
  constructor(private readonly target: string, private readonly replacement: string) {}

  apply(input: string): string {
    return input.split(this.target).join(this.replacement);
  }
}

File: string-transformer/python/src/string_transformer/pipeline.py
from dataclasses import dataclass
from typing import List, Protocol


# Byte-identical across C#, TypeScript, and Python. The truncation
# marker is part of the spec (see ../SCENARIOS.md).
TRUNCATION_MARKER = "..."


# Strategy. Every transformation is a self-contained rule — one input,
# one output, no state. Pipeline composes them; it does not know which
# rule any given strategy implements.
class Transformation(Protocol):
    def apply(self, input: str) -> str: ...


@dataclass(frozen=True)
class Pipeline:
    steps: List[Transformation]

    def run(self, input: str) -> str:
        current = input
        for step in self.steps:
            current = step.apply(current)
        return current


def _is_letter(c: str) -> bool:
    return ("a" <= c <= "z") or ("A" <= c <= "Z")


def _is_separator(c: str) -> bool:
    return c.isspace() or c == "-"


class Capitalise:
    def apply(self, input: str) -> str:
        if not input:
            return input
        chars = list(input)
        at_word_start = True
        for i, c in enumerate(chars):
            letter = _is_letter(c)
            if letter and at_word_start:
                chars[i] = c.upper()
            at_word_start = not letter
        return "".join(chars)


class Reverse:
    def apply(self, input: str) -> str:
        return input[::-1]


class RemoveWhitespace:
    def apply(self, input: str) -> str:
        return "".join(c for c in input if not c.isspace())


class SnakeCase:
    def apply(self, input: str) -> str:
        lowered = input.lower()
        out: List[str] = []
        in_separator = True  # suppress leading separators
        for c in lowered:
            if _is_separator(c):
                in_separator = True
            else:
                if in_separator and out:
                    out.append("_")
                out.append(c)
                in_separator = False
        return "".join(out)


class CamelCase:
    def apply(self, input: str) -> str:
        out: List[str] = []
        token_index = 0
        at_token_start = True
        for c in input:
            if _is_separator(c):
                if not at_token_start:
                    token_index += 1
                    at_token_start = True
            else:
                lower = c.lower()
                out.append(lower.upper() if at_token_start and token_index > 0 else lower)
                at_token_start = False
        return "".join(out)


@dataclass(frozen=True)
class Truncate:
    length: int

    def apply(self, input: str) -> str:
        if len(input) <= self.length:
            return input
        return input[: self.length] + TRUNCATION_MARKER


@dataclass(frozen=True)
class Repeat:
    times: int

    def apply(self, input: str) -> str:
        if self.times <= 0:
            return ""
        return " ".join([input] * self.times)


@dataclass(frozen=True)
class Replace:
    target: str
    replacement: str

    def apply(self, input: str) -> str:
        return input.replace(self.target, self.replacement)
